### PR TITLE
Add changesets and automated `npm` release

### PR DESCRIPTION
**Pull request by georgeblahblah**
## What problems is this designed to solve?

- Simpler version management using established tools
- Automated releases to `npm`
- Automated releases to a `snapshot` tag on `npm`, via Prereleases on GitHub Releases
- Easier to see a versioned changelog, which provides a history of changes

## What does this change?

- introduces [`changesets`](https://github.com/changesets/changesets) to the repository
- introduces a script (`./scripts/ci/update-version-sbt.sh`) to set the sbt build version correctly, based on the version defined in `package.json`. The version in `package.json` is managed by changesets, and is the source of truth
- introduces a GitHub Action to create a release PR, or publish unreleased changes for activity on `main`
  - this uses a custom publish script, `./scripts/ci/release-npm.sh` to build the models and publish using `sbt`
  - `./scripts/ci/release-snapshot-npm.sh` will release the models to the `snapshot` tag on `npm`. This means that `npm install @guardian/apps-rendering-api-models@latest` will ignore the release
- bumps the sbt plugin `scrooge-typescript-extras`, so we don't need a global typescript installation to exist

## How does automated publishing work?

There are two release modes: snapshot and regular. Snapshot releases are for testing changes, and will be ignored by `npm install @guardian/apps-rendering-api-models@latest`.

### Publishing a regular release

Regular releases are for tested changes you wish to be generally available.

- Once your feature is developed on a feature branch, add a changeset summarising the changes by running `npx changeset` locally
- Commit and push this changeset to your branch, and merge to `main` as normal
- Changesets will create or update a "Release PR", which includes all unreleased changes
- Merging this "Release PR" to `main` will publish the package to `npm`

### Publishing a snapshot release

- Make any changes you wish to test in a branch and push to GitHub
- Create a **Prerelease** in GitHub releases
  - Specify the "target" as your feature branch
  - Specify a unique tag, e.g `v0.0.0-2022-10-20-SNAPSHOT`
- Publish to prerelease to automatically publish the package to the `snapshot` tag on `npm`
- Install the package with `npm install @guardian/apps-rendering-api-models@snapshot` or `npm install @guardian/apps-rendering-api-models@v0.0.0-2022-1-20-SNAPSHOT`

## Why isn't this as simple as just running `npm publish`?

Because this repository itself doesn't contain the package to be published, but rather the source files from which a package is generated, there are various steps in `sbt` to generate the publish package from the source thrift files.

## Why changesets?

Changesets brings significantly easier version management. Rather than manually setting versions and remembering to push/commit those version changes, we can create changesets which set the version automatically. However, because we are not using changeset's publishing features, we don't get the full benefit of GitHub Releases being created automatically. However, I think the automated version setting and changelog management are significant benefits.

However, there is a small cost to changesets. We have to maintain a shell `npm` project at the root which exists only to run changesets from. Additionally, there is some developer overhead required to learn the relevant changeset commands.

## What other options were explored?

In addition to changesets, I explored a few other workflows:

**Using the [`dynver` sbt plugin](https://github.com/sbt/sbt-dynver)**

`dynver` sets he build version based on proximity of `HEAD` to git tags. I found the version resolution algorithm to be hard to understand. I also found that it does not easily support snapshot versioning, which is important for a package like `apps-rendering-api-models`.

**Managing the version manually**

This is pretty much the process we have now. It works, and could continue to work. However it adds some confusion around what the version should be.

## How to test

- Does the `npm` package publish automatically when merging the "Release PRs" or publishing a Prerelease on GitHub Releases?

### CODE REVIEW

**Comment by georgeblahblah** [comment]
When writing the documentation for the snapshot release process, I realised it might be needlessly complex. I'm doing a quick test locally to see if I can make this step easier to reason about.

File: README.md [hunk 1]
```diff
@@ -2,7 +2,27 @@
 
 These models are used to communicate with the Apps Rendering API
 
-## How to release
+## How to release to NPM
+
+This repository uses [`changesets`](https://github.com/changesets/changesets) to manage versions and releases.
+
+To create a changeset, ensure you are using the correct Node version defined in [`.nvmrc`](./.nvmrc) and project dependencies installed by running `npm install`. Then:
```
Comment (JamieB-gu) [inline README.md:9]: Should we mention that people should also use the `npm` version that corresponds to this Node version?

File: scripts/ci/release-npm.sh [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+pushd $(dirname $0)
+VERSION=$(./get-version.sh)
+popd
+
+if npm view "@guardian/apps-rendering-api-models@$VERSION" > /dev/null 2>&1 ; then
+    echo "$VERSION already released. Exiting"
+    exit 0
+else
+    echo "$VERSION not yet released. Releasing"
+    # sbt "project tsApiModels" "releaseNpm $VERSION"
```
Comment (JamieB-gu) [inline scripts/ci/release-npm.sh:14]: Will this publish while this is commented out?

**Review by JamieB-gu — COMMENTED** [review]
This is great! The docs are very clear. Just a few small questions.

File: .github/workflows/changesets.yaml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  create_release_pr:
+    name: Create release PR
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+
+      - name: Setup Node
+        uses: actions/setup-node@v3
+        with:
+          node-version-file: '.nvmrc'
```
Comment (JamieB-gu) [inline .github/workflows/changesets.yaml:18]: Worth adding a `cache` too?

File: .github/workflows/release-snapshot.yaml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    runs-on: ubuntu-latest
+    if: "github.event.release.prerelease"
+
+    steps:
+      - uses: actions/checkout@v3
+
+      - name: Setup Node
+        uses: actions/setup-node@v3
+        with:
+          node-version-file: '.nvmrc'
```
Comment (JamieB-gu) [inline .github/workflows/release-snapshot.yaml:19]: Worth adding a `cache` too?

File: README.md [hunk 5]
```diff
@@ -2,7 +2,27 @@
...
+## How to release a Snapshot to NPM
+
+- Create a branch which has the changes you want to test
+- Create a **Prerelease** using GitHub Releases
+    - Set the Target to your branch
+    - You must also create a tag for the snapshot release. For example, `v0.0.0-2022-10-20-SNAPSHOT`
+    - To automatically release the snapshot to `npm`, publish the prerelease
+    - Snapshots are released to the `snapshot` tag on `npm`. You can install them with `npm install @guardian/apps-rendering-api-models@snapshot`
+
+## How to release to NPM and Sonatype (legacy)
```
[the remark this answers is not in the document]
Comment (georgeblahblah) [reply README.md:25]: Good question. There is a chance they could become unsynchronised. I will remove the `npm` release instructions and leave the Sonatype release instructions.

Trying to release the same version twice will result in a failure on the second attempt (assuming the first attempt was successful), but it's still worth taking out the `npm` docs from here.

File: scripts/ci/release-npm.sh (hunk 2 above)
Comment (georgeblahblah) [reply scripts/ci/release-npm.sh:14]: 😱 no! I'll uncomment 😅

**Review by JamieB-gu — APPROVED** [review]

**Pull request merged**